### PR TITLE
[AIRFLOW-5522] BQ list dataset tables operator

### PR DESCRIPTION
**Pull request by mohannadbanayosi**
### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW-5522) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-5522

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

This operator will fetch the tables of the specified dataset. Currently it will return the entire BQ return object ([reference](https://cloud.google.com/bigquery/docs/reference/rest/v2/tables/list#response-body)) but we might think of extracting fewer parts.

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### CODE REVIEW

File: airflow/gcp/hooks/bigquery.py [hunk 1]
```diff
@@ -1430,6 +1430,38 @@ def cancel_query(self) -> None:
                               self.running_job_id)
                 time.sleep(5)
 
+    def get_dataset_tables(self, dataset_id: str, project_id: Optional[str] = None,
+                           max_results: Optional[int] = None, page_token: Optional[str] = None) -> Dict:
```
Comment (turbaszek) [inline airflow/gcp/hooks/bigquery.py]: Can we make the return type more precise? For example `Dict[str, Dict]`.

**Review by turbaszek — APPROVED** [review]
LGTM! Would like to extend the example DAG for BQ with your new operator?

**Comment by mohannadbanayosi** [comment]
Extended the BQ example DAG to include the new operator. @nuclearpinguin

File: airflow/gcp/hooks/bigquery.py (hunk 1 above)
Comment (mohannadbanayosi) [reply airflow/gcp/hooks/bigquery.py]: Changed it to `Dict[str, Union[str, int, List]]`.

File: airflow/gcp/hooks/bigquery.py [hunk 2]
```diff
@@ -1430,6 +1430,39 @@ def cancel_query(self) -> None:
                               self.running_job_id)
                 time.sleep(5)
 
+    def get_dataset_tables(self, dataset_id: str, project_id: Optional[str] = None,
+                           max_results: Optional[int] = None,
+                           page_token: Optional[str] = None) -> Dict[str, Union[str, int, List]]:
+        """
+        Get the list of tables for a given dataset.
+        see https://cloud.google.com/bigquery/docs/reference/rest/v2/tables/list
```
Comment (feluelle) [inline airflow/gcp/hooks/bigquery.py]:
```suggestion

        .. seealso:: https://cloud.google.com/bigquery/docs/reference/rest/v2/tables/list
```
Looks better imo. Don't you think so? :)

https://www.sphinx-doc.org/en/1.5/markup/para.html#directive-seealso

It creates colored boxes: https://thomas-cokelaer.info/tutorials/sphinx/rest_syntax.html#colored-boxes-note-seealso-todo-and-warnings

File: airflow/gcp/operators/bigquery.py [hunk 3]
```diff
@@ -1132,6 +1132,65 @@ def execute(self, context):
...
+    :type page_token: str
+    :param gcp_conn_id: (Optional) The connection ID used to connect to Google Cloud Platform.
+    :type gcp_conn_id: str
+    :param delegate_to: (Optional) The account to impersonate, if any.
+        For this to work, the service account making the request must have domain-wide
+        delegation enabled.
+    :type delegate_to: str
+
+    :rtype: dict
+        https://cloud.google.com/bigquery/docs/reference/rest/v2/tables/list#response-body
```
Comment (feluelle) [inline airflow/gcp/operators/bigquery.py]:
```suggestion

        .. seealso:: https://cloud.google.com/bigquery/docs/reference/rest/v2/tables/list#response-body
```

File: tests/gcp/operators/test_bigquery.py [hunk 4]
```diff
@@ -566,3 +566,26 @@ def test_execute(self, mock_hook):
...
+    @mock.patch('airflow.gcp.operators.bigquery.BigQueryHook')
+    def test_execute(self, mock_hook):
+        operator = BigQueryGetDatasetTablesOperator(
+            task_id=TASK_ID,
+            dataset_id=TEST_DATASET,
+            project_id=TEST_GCP_PROJECT_ID,
+            max_results=2
+        )
+
+        operator.execute(None)
```
Comment (feluelle) [inline tests/gcp/operators/test_bigquery.py:581]:
```suggestion
        operator.execute(context=None)
```
just my preference - change it only if you agree :)

File: airflow/gcp/hooks/bigquery.py [hunk 5]
```diff
@@ -1430,6 +1430,39 @@ def cancel_query(self) -> None:
...
+
+        :return: map containing the list of tables + metadata.
+        """
+        optional_params = {}
+        if max_results:
+            optional_params['maxResults'] = max_results
+        if page_token:
+            optional_params['pageToken'] = page_token
+
+        dataset_project_id = project_id if project_id else self.project_id
```
Comment (feluelle) [inline airflow/gcp/hooks/bigquery.py]:
```suggestion
        dataset_project_id = project_id or self.project_id
```
See: https://docs.python.org/3/library/stdtypes.html#truth-value-testing

**Review by feluelle — CHANGES_REQUESTED** [review]

File: airflow/gcp/hooks/bigquery.py (hunk 2 above)
Comment (mohannadbanayosi) [reply airflow/gcp/hooks/bigquery.py]: Totally agreed. 👍

**Comment by mohannadbanayosi** [comment]
Done with the requested changes. :-) @feluelle

**Review by feluelle — APPROVED** [review]
LGTM

**Pull request merged**